### ISSUE TITLE
Custom parsers

### ISSUE THREAD

**Issue by ai** [postcss/postcss#140]
1. Someone want to parser SCSS with `// one line comment`.
2. I want to have light syntax without `;` (and maybe with one line comment).
3. Some legacy code requires Safe Mode parser to forget big CSS issues. Not it is in parser by options and it is not good solution.

Maybe we should allow PostCSS to change parsers:

```js
postcss.process(css);
postcss.process(css, { parser: SafeParser });
```

All parser should return same AST, but them can parse any syntax.

/cc @lydell @MoOx @rvanes

**Comment by onemanstartup** [comment]
```sass
.AppMenubar
  box-shadow 0px 2px 4px 0px rgba(0,0,0,0.07)
  a
    color #838383
    border-bottom 4px solid transparent
    &.active, &:hover
      border-bottom 4px solid #D8D8D8
```

Will it be possible? I want to be able parse this :)

**Comment by ai** [comment]
@onemanstartup you can do it right now with [css-whitespace](https://github.com/reworkcss/css-whitespace) compiler (it compiles this syntax to CSS, CSS goes to PostCSS).

**Comment by ai** [comment]
@onemanstartup I think about different syntax, because multiline selectors and values is important (for example, for multiple gradients). So we will have `{`, but not `;` in my current idea.

**Comment by onemanstartup** [comment]
@ai stylus support multiline selectors
```sass
h1, 
h2
  color #23303C
```
So css-whitespace doesn't support this? Then I'm sticking with stylus for now :) 

```sass
   background-image: radial-gradient(closest-corner at 40% 70%,#FFFFFF 0%, rgb(171,171,171),50%,transparent),
                     radial-gradient(closest-corner circle at 80% 20%, #FFFFFF 0%, rgb(171,171,171),20%,transparent),
```
Why do you need {}? multiline have commas as delimeters

**Comment by ai** [comment]
@onemanstartup hm, nice. It allows `\n` after `,`?

**Comment by onemanstartup** [comment]
@ai eh, do you mean `\n\n`? or as I show 
```
h1,
h2
```
It supports
```
.public_fixedDataTableRow_main

,
.public_fixedDataTableCell_wrap1

.public_fixedDataTableRow_main2
  overflow auto
```
```
.public_fixedDataTableRow_main
.public_fixedDataTableCell_wrap1
.public_fixedDataTableRow_main2
  overflow auto
```
```
.public_fixedDataTableRow_main

,.public_fixedDataTableCell_wrap1
,.public_fixedDataTableRow_main2
  overflow auto
```
It's all becomes
```
.public_fixedDataTableRow_main,
.public_fixedDataTableCell_wrap1,
.public_fixedDataTableRow_main2 {
  overflow: auto;
}
```

**Comment by lydell** [comment]
(Just a little note about css-whitespace: it does not support source maps).

**Comment by corysimmons** [comment]
Please God kill the curly braces. :|

**Comment by tunnckoCore** [comment]
I think the first thing to that way would be to externalize the parser. And definitely should be for v5 milestone.

**Comment by ai** [comment]
@tunnckoCore custom parsers will be sooner in 4.2.

**Comment by ai** [comment]
I think about this API:

```js
postcss.process(scss, { parser: 'scss' }).css //=> CSS output
postcss.process(scss, { syntax: 'scss' }).css //=> SCSS output for CSSComb for example
postcss.process(css, { stringifier: 'scss' }).css //=> SCSS output from CSS
```

So we need to rename `Result#css`. `Result#content`?

**Comment by ai** [comment]
Also:

```js
postcss.process(scss, { parser: 'scss' }) // Will load require('postcss-scss-syntax').parser
postcss.process(scss, { parser: require('custom-naming').parser })
```

**Comment by tunnckoCore** [comment]
> So we need to rename Result#css. Result#content?

yep and yep.

**Comment by ben-eb** [comment]
:+1: for `contents`.

**Comment by lydell** [comment]
I'd vote for `code`.

**Comment by MoOx** [comment]
`Result#output` ?
`contents` looks fine too.

**Comment by pkyeck** [comment]
so, single line comments are possible right now? or with the upcoming 5.0 release?

**Comment by ai** [comment]
Right now we have only hack without selector support

**Comment by lydell** [comment]
@pkyeck There’s also https://github.com/lydell/strip-css-singleline-comments.

**Comment by ai** [comment]
чт, 21 мая 2015, 13:46, Simon Lydell <notifications@github.com>:

**Comment by yisibl** [comment]
Inline comment +1

**Comment by ai** [comment]
Work is almost finished in [valac](https://github.com/postcss/postcss/tree/valac) branch.

Ready:

* All stringify code was moved from nodes classes to one Stringifier class.
* Stringifier was refactored to be more extendable.
* Options `parser`, `stringifier` and `syntax` was added to `Processor#process`.

Need to do:

* Write a docs how to wrtie own syntax.

**Comment by bfred-it** [comment]
Does this mean one will be able to run PostCSS on SCSS files?

Currently I run PostCSS after Sass but that limits its functionality considerably when Sass encounters syntax it doesn't understand. With this I can run PostCSS **before** Sass?

**Comment by ai** [comment]
@bfred-it yeap, you will be able to use PostCSS tranformations on SCSS files (of course, PostCSS will not be able to *compile* SCSS to CSS, not be able to expand mixins, etc).

**Comment by ai** [comment]
Oops, this bug is blocking 4.2 npm/npm#2063

`postcss` depends on `postcss-safe-parser` to be able to emulate depreacted `safe` option. But `postcss-safe-parser` depends on `postcss` to extend default parser.

Solutions:

* Split `postcss` into `postcss-ast`, `postcss-css`, etc. I am not ready for this big refactoring with many places to make mistakes.
* Release 4.2 as 5.0, so we can remove `safe` option. Anyway it is used by rare users. Of course, most of “current 5.0” refactoring will be moved to 6.0.

What do you think? Do we ready for major bump? Or maybe we have other solution?

**Comment by ben-eb** [comment]
I don't use the `safe` option at all so I would be happy with removing it and releasing a major version.

**Comment by tunnckoCore** [comment]
@ai split core out.

Im sure devs would love the idea to have separate `postcss-parser`, `postcss-compilier` and etc.

**Comment by ai** [comment]
@tunnckoCore I like a splitting too, but I am afraid that it will delay 5.0 for a month :(. Also we need more strange user cases (like using PostCSS in React Style) to split it in right way.

**Comment by lydell** [comment]
Let’s not forget to learn from history. There used to be `css-parse` and `css-stringify` used by rework, but those were merged into `css` because it was too difficult to maintain otherwise. Just sayin’.

**Comment by ai** [comment]
@lydell good note.

**Comment by ai** [comment]
Oups, if next release will be 5..0, there will be 2 broken changes:

1. `safe` options was removed.
2. `Node#before` and other raw properties were moved to `Node#raw.before`.

I can make a more clear errors to migrate plugins easily.

@MoOx what do you think about major release?

**Comment by MoOx** [comment]
major is always safer than minor :)

Side note: FYI, I will probably work with (inline) styles with JS in the near future so I won't be as active as before in postcss ecosystem.

**Comment by ai** [comment]
@MoOx :'(

**Comment by MoOx** [comment]
I think CSS is definitly not the future. The first letter is clearly one big problem.

**Comment by iamstarkov** [comment]
@MoOx is cssnext in danger? =)

**Comment by davidtheclark** [comment]
@MoOx: Too bad. I'm yet to be convinced by the style-in-JS advocates, mainly because on my current big project CSS works great, and trying to do the same things in JS would be a big headache, lots of added complexity. We will see! Curious to hear about it.

**Comment by MoOx** [comment]
I am writing React based UI for 2 years now and I don't see how I can go back to anything else. CSS bring some issues like vjeux exposed. So I will work on a new project to bring in JS some convenient way to write some styles.

That's said, I still have some Wordpress to handle, so, no cssnext is not in danger yet (+ reminder: it's open source and based on small modules, so nobody should worry about cssnext future).

**Comment by ai** [comment]
@MoOx BTW, I think cascade is a problem too. This is why I use only BEM or CSS Modules.

**Comment by safareli** [comment]
@MoOx  recently i was watching [React: CSS in JS ](http://blog.vjeux.com/2014/javascript/react-css-in-js-nationjs.html) and then came to [this document](https://github.com/gilbox/css-bliss/blob/master/solving-complexity.md) about solving Facebook's CSS complexity problems with [CSS Bliss](https://github.com/gilbox/css-bliss) so definitely check it if you haven't done already

**Comment by ai** [comment]
SCSS parser is ready: https://github.com/postcss/postcss-scss

Safe parser was moved to separated project: https://github.com/postcss/postcss-safe-parser

Only syntax docs are left.

**Comment by ai** [comment]
Syntax docs are finished https://github.com/postcss/postcss/blob/valac/docs/syntax.md

**Comment by varya** [comment]
Oh, sooo nice to see it finished. Any plans of havign parsers for SASS, LESS, SCSS? And, if they appear, with they be in separate repositories?

**Comment by andyjansson** [comment]
@varya SCSS parser already exists: https://github.com/postcss/postcss-scss

**Comment by ai** [comment]
@varya right now I have no plans for Stylus, Less abd Sass support (just have no time). This is why I wrote a good docs about implementing syntax:

https://github.com/postcss/postcss/blob/valac/docs/syntax.md

**Comment by ai** [comment]
Please write what you need from next Styles/Sass like syntax: #495

**Comment by jamiehutber** [comment]
Did this ever get released? Because I still can't use comments

````
ERROR in ./~/css-loader?modules=true&importLoaders=1&localIdentName=%5Bpath%5D%5Bname%5D-%5Blocal%5D!./~/postcss-loader!./src/client/scss/main.css
C:\var\www\webpack-express-boilerplate\src\client\scss\main.css:1:1: Unknown word
// Immutable items
^`````

**Comment by ai** [comment]
@jamiehutber you have issue, because you didn’t *change* parser ;)

**Comment by jamiehutber** [comment]
Dam you're good on the whole reply thing 🌴 

Nice one, I should have read the docs for postcss-scss properly :)

**Comment by GoudekettingRM** [comment]
@ai I'm sure I'm missing something, but I can't seem to get it to work:

```js
// postcss.config.js
module.exports = {
  parser: 'postcss-scss',
  plugins: [
    require('postcss-import-ext-glob'),
    require('postcss-import'),
    require('tailwindcss/nesting'),
    require('tailwindcss'),
    require('postcss-flexbugs-fixes'),
    require('autoprefixer'),
    require('postcss-preset-env')({
      autoprefixer: {
        flexbox: 'no-2009',
      },
      stage: 3,
    }),
  ],
};
```

error: 
```
34:32 css.1  | CssSyntaxError: postcss-import: app/assets/stylesheets/auctions.scss:1:4: Unknown word
17:34:32 css.1  |
17:34:32 css.1  | > 1 | // Place all the styles related to the auctions controller here.
17:34:32 css.1  |     |    ^
```

Versions:
```
"postcss": "^8.4.14",
"postcss-scss": "^4.0.4",
"postcss-import-ext-glob": "^2.0.1",
"postcss-flexbugs-fixes": "^5.0.2",
"postcss-import": "^14.1.0",
"postcss-preset-env": "^7.7.0",
```


I also tried postcss-sass and I tried putting it for on the syntax property, but no dice.
